File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.java
import java.util.*;

class Solution {
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        Queue<Integer> bridge = new LinkedList<>();
        int time = 0, totalWeight = 0, idx = 0;
        for(int i = 0; i < bridge_length; i++) bridge.add(0);
        
        
        while(idx < truck_weights.length){
            time++;
            totalWeight -= bridge.poll();
            
            if(totalWeight + truck_weights[idx] <= weight){
                bridge.add(truck_weights[idx]);
                totalWeight += truck_weights[idx];
                idx++;                
            }else{
                bridge.add(0);
            }
            
        }
        return time + bridge_length;
        
        
    }
}

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.kt
import java.util.*;
class Solution {
   class Truck(var weight: Int) {
        var move = 1
    }

    fun solution(bridge_length: Int, weight: Int, truck_weights: IntArray): Int {
        var answer = 0
        var waitQue: Queue<Truck> = LinkedList()
        var moveQue: Queue<Truck> = LinkedList()
        var curr_weight = 0

        for (i in truck_weights) {
            waitQue.offer(Truck(i))
        }

        while (moveQue.isNotEmpty() || waitQue.isNotEmpty()) {
            answer++

            if(moveQue.isEmpty()) {
                var t = waitQue.poll()
                curr_weight += t.weight
                moveQue.offer(t)
                continue
            }

            moveQue.forEach { it.move++ }

            if(moveQue.peek().move > bridge_length) {
                curr_weight -= moveQue.poll().weight
            }

            if(waitQue.isNotEmpty() && waitQue.peek().weight + curr_weight <= weight) {
                var t = waitQue.poll()
                curr_weight += t.weight
                moveQue.offer(t)
            }
        }

        return answer
    }
}